### ISSUE TITLE
[no-unnecessary-type-assertion] Misses "const x: object = {} as object"

### ISSUE THREAD

**Issue by ilyub** [typescript-eslint/typescript-eslint#3723]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "warn"
  }
}
```

```TS
export function fromEntries(): object {
  const result: object = {} as object;
  ...
  return result as object;
}
```

**Expected Result**

I think the rule should report "const" and "return" lines.

**Actual Result**

It reports only "return" line.

**Additional Info**

This is related to https://github.com/typescript-eslint/typescript-eslint/issues/3722

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.29.1` |

**Comment by bradzacher** [comment]
The `const` line is not covered by the rule - the rule only covers `as` assertions.
`{} as object` is necessary, as `{}` is a unique type that is distinct from `object`.

The return statement is covered by the rule:
![image](https://user-images.githubusercontent.com/7462525/128828070-16f7e78e-90c1-4932-b5c8-2a288dcbf12c.png)
